pkg/storage/stores/shipper/compactor/deletion/gennumber_loader.go

```diff
@@ -0,0 +1,173 @@
+package deletion
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/go-kit/log/level"
+
+	"github.com/grafana/loki/pkg/util/log"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const reloadDuration = 5 * time.Minute
+
+type GenNumberLoader struct {
+	numberGetter genNumberGetter
+	numbers      map[string]string
+	quit         chan struct{}
+	lock         sync.RWMutex
+	metrics      *genLoaderMetrics
+}
+
+type genNumberGetter interface {
+	GetCacheGenerationNumber(ctx context.Context, userID string) (string, error)
+}
+
+type genLoaderMetrics struct {
+	cacheGenLoadFailures prometheus.Counter
+}
+
+func NewGenNumberLoader(g genNumberGetter, registerer prometheus.Registerer) *GenNumberLoader {
+	if g == nil {
+		g = &noopNumberGetter{}
+	}
+
+	l := &GenNumberLoader{
+		numberGetter: g,
+		numbers:      make(map[string]string),
+		metrics:      newGenLoaderMetrics(registerer),
+	}
+	go l.loop()
+
+	return l
+}
+
+func newGenLoaderMetrics(r prometheus.Registerer) *genLoaderMetrics {
+	return &genLoaderMetrics{
+		cacheGenLoadFailures: promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "delete_cache_gen_load_failures_total",
+			Help:      "Total number of failures while loading cache generation number using gen number loader",
+		}),
+	}
+}
+
+func (l *GenNumberLoader) loop() {
+	timer := time.NewTicker(reloadDuration)
+	for {
+		select {
+		case <-timer.C:
+			err := l.reload()
+			if err != nil {
+				level.Error(log.Logger).Log("msg", "error reloading tombstones", "err", err)
```

Comment:
Tombstones doesn't look like the right term here. Can we use cache generation numbers or something else?

Reply:
Good call. I pulled a lot of this from Cortex and missed this